### PR TITLE
interpreter: (cf) add cf interpreter functions and extend arith

### PR DESCRIPTION
**Pull request by superlopuh**

### CODE REVIEW

File: tests/interpreters/test_cf_interpreter.py [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+from xdsl.dialects import arith, cf, func
+from xdsl.dialects.builtin import ModuleOp, i32
+from xdsl.interpreter import Interpreter
+from xdsl.interpreters.arith import ArithFunctions
+from xdsl.interpreters.cf import CfFunctions
+from xdsl.interpreters.func import FuncFunctions
+from xdsl.ir.core import Block, Region
+
+
+def triangle_fn(n: int) -> int:
```
Comment (webmiche) [inline tests/interpreters/test_cf_interpreter.py:11]: Where does this name vome from? Isn't this just the sum from 1 to n? 😅

File: tests/interpreters/test_cf_interpreter.py [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+    (result,) = interpreter.call_op("triangle", (n,))
+    return result
+
+
+def test_triangle():
+    assert triangle_fn(0) == triangle_interp(0)
+    assert triangle_fn(1) == 1
+    assert triangle_fn(2) == 3
+    assert triangle_fn(3) == 6
+    assert triangle_fn(4) == 10
```
Comment (webmiche) [inline tests/interpreters/test_cf_interpreter.py:69]: Why are you mainly testing the python function here? Why not make this a parametric test taking in a range of values and comparing the two results?

File: xdsl/interpreter.py [hunk 3]
```diff
@@ -362,25 +362,6 @@ def run_block(self, block: Block, args: PythonValues) -> PythonValues | None:
         The terminator of this block is expected either to call its successor or return
         the results for the region directly.
         """
```
Comment (webmiche) [inline xdsl/interpreter.py:364]: Is this git diffs messing with me or is this function now just empty? 🤔

File: xdsl/interpreters/arith.py
Comment (webmiche) [inline xdsl/interpreters/arith.py:1]: I guess we see here one of the limitations of the Pyhton-based interpreter: Modellling fixed-size integers is nasty in Python. That's also why signed an unsigned are not distinguished in Python...

But I guess as long as we stay within reasonable boundaries, this should not matter.

File: xdsl/interpreters/cf.py [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+class CfFunctions(InterpreterFunctions):
+    @impl_terminator(cf.Branch)
+    def run_br(self, interpreter: Interpreter, op: cf.Branch, args: tuple[Any, ...]):
+        return Successor(op.successor, args), ()
+
+    @impl_terminator(cf.ConditionalBranch)
+    def run_cond_br(
+        self, interpreter: Interpreter, op: cf.ConditionalBranch, args: tuple[Any, ...]
+    ):
+        cond: int = args[0]
```
Comment (webmiche) [inline xdsl/interpreters/cf.py:23]:
```suggestion
        cond: bool = args[0]
```

Why not like this?

Comment (compor) [reply xdsl/interpreters/cf.py:23]: condition is `i1` in cf, but I guess in Python-land we can cast it?

**Review by compor — APPROVED** [review]

**Review by adutilleul — APPROVED** [review]

File: tests/interpreters/test_cf_interpreter.py (hunk 1 above)
Comment (superlopuh) [reply tests/interpreters/test_cf_interpreter.py:11]: In my head it made sense :) it's the number of elements in a lower triangular matrix of size n. I'll rename to `sum_to`

File: xdsl/interpreters/cf.py (hunk 4 above)
Comment (superlopuh) [reply xdsl/interpreters/cf.py:23]: fun fact :)
```
>>> issubclass(bool, int)
True
```

File: tests/interpreters/test_cf_interpreter.py (hunk 2 above)
Comment (superlopuh) [reply tests/interpreters/test_cf_interpreter.py:69]: I just forgot to finish writing this bit of code, will do the parametric thing

File: xdsl/interpreters/cf.py (hunk 4 above)
Comment (superlopuh) [reply xdsl/interpreters/cf.py:23]: So it really depends on whether you philosophically think of i1 as bool or int, at runtime it's still going to be an instance of bool there's no actual casting going on

File: xdsl/interpreter.py (hunk 3 above)
Comment (superlopuh) [reply xdsl/interpreter.py:364]: Hmm I'll check the whole function should be deleted

File: tests/interpreters/test_cf_interpreter.py (hunk 1 above)
Comment (webmiche) [reply tests/interpreters/test_cf_interpreter.py:11]: I see, that's true :sweat_smile:

File: xdsl/interpreters/cf.py (hunk 4 above)
Comment (webmiche) [reply xdsl/interpreters/cf.py:23]: Ah okay, that's fine then.

File: xdsl/interpreters/arith.py
Comment (math-fehr) [reply xdsl/interpreters/arith.py:1]: Yeah, I guess the problem here is that our interpreter should also take into account bitwidth, which it doesn't now!

**Review by math-fehr — APPROVED** [review]
Nice!

File: tests/interpreters/test_cf_interpreter.py [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+from xdsl.interpreters.arith import ArithFunctions
+from xdsl.interpreters.cf import CfFunctions
+from xdsl.interpreters.func import FuncFunctions
+from xdsl.ir.core import Block, Region
+
+
+def triangle_fn(n: int) -> int:
+    result = 0
+    i = 0
+    while i <= n:
```
Comment (georgebisbas) [inline tests/interpreters/test_cf_interpreter.py:14]: This can be done better?

Comment (superlopuh) [reply tests/interpreters/test_cf_interpreter.py:14]: Is that a question or an assertion? the whole function is `n * (n-1) // 2` so can be done better indeed :)

**Review by webmiche — APPROVED** [review]

**Pull request merged**